apps/astarte_appengine_api/lib/astarte_appengine_api/queries.ex

```diff
@@ -31,38 +30,66 @@ defmodule Astarte.AppEngine.API.Queries do
         {:ok, pem}
 
       {:error, %Xandra.ConnectionError{} = err} ->
-        _ =
-          Logger.warning("Database connection error #{Exception.message(err)}.",
-            tag: "database_connection_error"
-          )
+        Logger.warning("Database connection error #{Exception.message(err)}.",
+          tag: "database_connection_error"
+        )
 
         {:error, :database_connection_error}
 
       {:error, %Xandra.Error{} = err} ->
         handle_xandra_error(err)
+
+      {:error, error} ->
+        {:error, error}
     end
   end
 
-  defp do_fetch_public_key(keyspace_name, conn) do
-    query = """
-    SELECT blobAsVarchar(value)
-    FROM #{keyspace_name}.kv_store
-    WHERE group='auth' AND key='jwt_public_key_pem';
-    """
-
-    with {:ok, prepared} <- Xandra.prepare(conn, query),
-         {:ok, page} <-
-           Xandra.execute(conn, prepared, %{},
-             uuid_format: :binary,
-             consistency: :quorum
-           ) do
-      case Enum.to_list(page) do
-        [%{"system.blobasvarchar(value)" => pem}] ->
-          {:ok, pem}
-
-        [] ->
-          {:error, :public_key_not_found}
-      end
+  defp do_fetch_public_key(keyspace_name, _conn) do
+    schema_query =
+      from r in KvStore,
+        prefix: ^keyspace_name,
+        select: fragment("blobAsVarchar(?)", r.value),
+        where: r.group == "auth" and r.key == "jwt_public_key_pem"
+
+    opts = [uuid_format: :binary, consistency: :quorum]
+
+    case safe_query(schema_query, opts) do
+      {:ok, %{rows: [[public_key]]}} ->
+        {:ok, public_key}
+
+      {:ok, %{num_rows: 0}} ->
+        Logger.warning("No public key found in realm #{keyspace_name}.",
+          tag: "no_public_key_found"
+        )
+
+        {:error, :public_key_not_found}
+
+      {:error, error} ->
+        {:error, error}
+    end
+  end
+
+  defp safe_query(ecto_query, opts) do
+    {sql, params} = Repo.to_sql(:all, ecto_query)
+
+    # Equivalent to a `Repo.all`, but does not raise if we get a Xandra Error.
+    case Repo.query(sql, params, opts) do
+      {:ok, result} ->
+        {:ok, result}
+
+      {:error, %Xandra.ConnectionError{}} ->
+        {:error, :database_connection_error}
+
+      {:error, %Xandra.Error{message: message, reason: reason}} ->
+        error_message =
+          case message do
+            "" -> inspect(reason)
+            message -> message
+          end
+
+        Logger.warning("Health is not good: #{error_message}", tag: "db_health_check_bad")
+
+        {:error, :health_check_bad}
```

Comment:
the definition you have now is for health checks

```suggestion
      {:error, %Xandra.Error{} = xandra_error} ->
        handle_xandra_error(xandra_error)
```